File: src/day1.py
import utils

def __main__():
    lines = utils.read_input()
    nums = [int(line) for line in lines]
    target = 2020

    print(two_sum(nums, target))

def two_sum(nums: list[int], target: int) -> int:
    seen = set()
    for num in nums:
        needed = target - num
        if needed in seen:
            return num * needed
        seen.add(num)

__main__()

File: rust/src/day1.rs
mod utils;

fn main() {
    let lines = utils::read_input();
    let nums: Vec<i32> = lines
        .iter()
        .map(|line| line.parse().expect("Failed to parse number"))
        .collect();

    let target = 2020;

    println!("{}", two_sum(&nums, target));
}

fn two_sum(nums: &[i32], target: i32) -> i32 {
    let mut seen = std::collections::HashSet::new();

    for &num in nums {
        let needed = target - num;
        if seen.contains(&needed) {
            return num * needed;
        }
        seen.insert(num);
    }

    panic!("No solution found")
}
